Implement readout protection for STM32H7

FRASTM opened the pull request:
following the https://github.com/zephyrproject-rtos/zephyr/pull/76097

This one is adding the READOUT protection (RDP) for the stm32h7 serie

Tested on the stm32h745i disco with tests/drivers/flash/stm32/  with CONFIG_FLASH_STM32_READOUT_PROTECTION=y

```
*** Booting Zephyr OS build v3.7.0-449-ge7cd0f0307a2 ***                                                               
Running TESTSUITE flash_stm32                                                                                          
===================================================================                                                    
START - test_stm32_readout_protection_disabled                                                                         
RDP is disabled                                                                                                        
 PASS - test_stm32_readout_protection_disabled in 0.002 seconds                                                        
===================================================================                                                    
TESTSUITE flash_stm32 succeeded                                                                                        
                                                                                                                       
------ TESTSUITE SUMMARY START ------                                                                                  
                                                                                                                       
SUITE PASS - 100.00% [flash_stm32]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.002 seconds                   
 - PASS - [flash_stm32.test_stm32_readout_protection_disabled] duration = 0.002 seconds                                
                                                                                                                       
------ TESTSUITE SUMMARY END ------                                                                                    
                                                                                                                       
===================================================================                                                    
PROJECT EXECUTION SUCCESSFUL          
```

FRASTM commented:
Also etsed on the stm32h750b_dk to answer the https://github.com/zephyrproject-rtos/zephyr/discussions/76297

FRASTM commented on drivers/flash/flash_stm32.h:
```diff
@@ -295,6 +295,12 @@ int flash_stm32_option_bytes_lock(const struct device *dev, bool enable);
 
+#ifdef CONFIG_SOC_SERIES_STM32H7X
+#define flash_stm32_option_bytes_disable flash_stm32h7_option_bytes_disable
+int flash_stm32h7_option_bytes_disable(const struct device *dev);
```
> I'm in general confused by the introduction of the define one line above. Could you explain its purpose?

the purpose is to remap stm32h7 specific function and keep the  flash_stm32_ex_op.c also for the stm32H7 serie

benediktibk replied:
Why is it necessary to keep the other version available as well?

benediktibk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -90,6 +91,109 @@ static inline void _flash_stm32_sem_give(const struct device *dev)
+}
+#endif /* CONFIG_FLASH_STM32_READOUT_PROTECTION */
+
+int flash_stm32_option_bytes_lock(const struct device *dev, bool enable)
```
But this means that the implementation in flash_stm32.c doesn't work for the H7 series?
- If yes, then I think we should remove the comment there.
- If no, why is there another implementation for it?

FRASTM replied:
that part (flash_stm32.c) has not been updated here, removing the reference to the stm32h7 serie though

benediktibk replied:
Thanks, this makes it more clear.

duda-patryk commented:
@FRASTM Is there any progress on that? I'm interested in support for RDP and WP for STM32H7 chips

FRASTM commented on drivers/flash/flash_stm32_ex_op.c:
```diff
@@ -85,13 +88,79 @@ int flash_stm32_ex_op_sector_wp(const struct device *dev, const uintptr_t in,
+	target_level = current_level;
+
+	/*
+	 * 0xAA = RDP level 0 (no protection)
+	 * 0xCC = RDP level 2 (permanent protection)
+	 * others = RDP level 1 (protection active)
```
yes 3 level are defined for all the stm32 mcus
- RDP_LEVEL_0 No protection  --> 0xAA
- RDP_LEVEL_1 Read protection of the memory --> 0xBB
- RDP_LEVEL_2 Full chip protection --> 0xCC

duda-patryk replied:
Sounds good!

FRASTM commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -90,6 +91,109 @@ static inline void _flash_stm32_sem_give(const struct device *dev)
+}
+
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+static int flash_stm32h7_ex_op(const struct device *dev, uint16_t code,
```
In case of stm32h7, the flash_stm32.c is not compiled, only the flash_stm32h7x.c is, based on the SoC. As a flash_stm32h7_ex_op function is needed, it seems simplier to have it here

benediktibk replied:
I'm pretty sure that it is not a good idea to copy code. How about introducing another file, like *_common.c or *_utils.c which includes these things?

FRASTM replied:
Ok I can try to move flash_stm32_ex_op from the flash_stm32 to the flash_stm32_ex_op file. So that it will be common for stm32h7 too.

- Note that _flash_stm32_sem_take and _flash_stm32_sem_give functions are "duplicated" for stm32h7 or other stm32 devices

FRASTM committed:
drivers: flash: stm32 driver move sem functions for multithread

Place the flash_stm32_sem_take and flash_stm32_sem_give
function to header file for common to use in any flash driver

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 flash ex op functions are common

Move the stm32 flash driver ex_op functions to common
flash_stm32_ex_op.c file.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32h7 with RDP protection

Implement the readout protection for the STM32H7 series.
Define the specific functions used by the flash_stm32h7.
Move the stm32h7 flash register manipulation in a write_optb()
to be called by the flash_stm32_set_rdp_level.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: flash RDP testing on stm32h7

Add the configuration to run the testcase
with READOUT_PROTECTION (RDP) on stm32h7 disco kits

Signed-off-by: Francois Ramu <francois.ramu@st.com>

benediktibk commented:
> @benediktibk Thanks for the initial review on this. Would you mind resolving conversations on addressed points when you have some free cycles ?

Done, but I was only able to comment on the reviews where they were done. As I am neither the PR owner nor have write access to the repo github won't let me resolve my own conversations :-|

niedzwiecki-dawid commented:
Can I somehow help to get this merged? :-)

de-nordic commented:
> Can I somehow help to get this merged? :-)

Review it?

niedzwiecki-dawid commented:
> > Can I somehow help to get this merged? :-)
> 
> Review it?

Sure, I can do that, but there are still open comments from previous reviews. I'm wondering if there are any blockers.

FRASTM commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -90,6 +91,109 @@ static inline void _flash_stm32_sem_give(const struct device *dev)
 #define flash_stm32_sem_give(dev)
 #endif
 
+static __unused int write_optb(const struct device *dev, uint32_t mask,
```
in that case, I prefer to move the write_optb function for the stm32H7 _direclty_ in the core of the flash_stm32_set_rdp_level(),
considering that void function did not used the returned code of the former write_optb()
There is no need for adding a write_optb especially for the stm32H7 
(if we do not want to duplicate code changing the register names)

niedzwiecki-dawid commented:
Can `stm32_control_register_disable` and `flash_stm32_option_bytes_disable` be moved to a common file or implementation of these functions can differ depending on soc family?

benediktibk commented:
> Can `stm32_control_register_disable` and `flash_stm32_option_bytes_disable` be moved to a common file or implementation of these functions can differ depending on soc family?

Both of these functions are used in flash_stm32_ex_op.c, but flash_stm32.c is not built always when flash_stm32_ex_op.c is being built. The build succeeds for now as the usage of these functions is within a #ifdef, but in general this construct seems to be brittle and confusing.

duda-patryk commented:
> but in general this construct seems to be brittle and confusing.

Yeah, I agree this could be confusing, but essentially the intention is to expose implementation of the extended operation to `flash_stm32_ex_op.c`.

By the way, the intention of `flash_stm32_ex_op.c` was to keep high level extended operation logic and perform verification which is normally performed in `flash_handlers.c`. The intention of this file is not to provide low-level implementation (SoC dependent), but I'm okay with keeping functions like `flash_stm32_ex_op_update_rdp()` there, because all STM32 devices have the same concept of RDP.

I'm aware that current state of things is not perfect, but definitely this is something that works. If you have some ideas on improving this construct (or flash extended operations in general) please let us know.

benediktibk commented:
> > but in general this construct seems to be brittle and confusing.
> 
> Yeah, I agree this could be confusing, but essentially the intention is to expose implementation of the extended operation to `flash_stm32_ex_op.c`.
> 
> By the way, the intention of `flash_stm32_ex_op.c` was to keep high level extended operation logic and perform verification which is normally performed in `flash_handlers.c`. The intention of this file is not to provide low-level implementation (SoC dependent), but I'm okay with keeping functions like `flash_stm32_ex_op_update_rdp()` there, because all STM32 devices have the same concept of RDP.
> 
> I'm aware that current state of things is not perfect, but definitely this is something that works. If you have some ideas on improving this construct (or flash extended operations in general) please let us know.

I think the approach for STM32 clock control is pretty well designed. Having a common file (clock_stm32_ll_common.c), which is built always (when at least one STM32 clock controller is available) and a device specific file (clock_stm32c0.c).
Something similar would also suit this case and would be more straight forward than the dependencies of the files in the current state.

de-nordic commented:
Hey @FRASTM , moving forward with this?

FRASTM commented on samples/basic/blinky_pwm/boards/nucleo_f070rb.overlay:
by error, removing

FRASTM commented on drivers/flash/CMakeLists.txt:
```diff
@@ -75,8 +75,10 @@ endif()
 if(CONFIG_SOC_FLASH_STM32)
   if(CONFIG_SOC_SERIES_STM32H7X)
     zephyr_library_sources_ifdef(CONFIG_DT_HAS_ST_STM32H7_FLASH_CONTROLLER_ENABLED flash_stm32h7x.c)
+    zephyr_library_sources_ifdef(CONFIG_FLASH_EX_OP_ENABLED flash_stm32_ex_op.c)
```
not sure, the 
`(CONFIG_SOC_SERIES_STM32WBX)`
have not

duda-patryk replied:
`CONFIG_FLASH_EX_OP_ENABLED` still needs to be enabled manually and `CONFIG_FLASH_HAS_EX_OP` must be selected, so I think it can be safely moved under `CONFIG_SOC_FLASH_STM32`.

FRASTM commented on drivers/flash/flash_stm32.c:
```diff
@@ -319,7 +292,7 @@ int flash_stm32_option_bytes_lock(const struct device *dev, bool enable)
 {
 	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
 
-#if defined(FLASH_OPTCR_OPTLOCK) /* F2, F4, F7 and H7 */
```
ok

duda-patryk replied:
I still see it in patch that moves sem functions.

duda-patryk replied:
This comment was not addressed

FRASTM commented on drivers/flash/flash_stm32_ex_op.c:
```diff
@@ -20,6 +20,46 @@
 
 LOG_MODULE_REGISTER(flash_stm32_ex_op, CONFIG_FLASH_LOG_LEVEL);
 
+extern int flash_stm32_ex_op_sector_wp(const struct device *dev, const uintptr_t in,
```
done

FRASTM commented on drivers/flash/flash_stm32.h:
```diff
@@ -351,14 +351,16 @@ uint8_t flash_stm32_get_rdp_level(const struct device *dev);
+int flash_stm32_option_bytes_disable(const struct device *dev);
+#endif
+
 /* Flash extended operations */
```
flash_stm32_ex_op_sector_wp() removed from flash_stm32.h

FRASTM commented:
rebased on 03959a20f7834ffe50a51df2b2e45ef87d1b610b
takes @duda-patryk remarks

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -23,7 +23,8 @@
 #include "flash_stm32.h"
-#include "stm32_hsem.h"
+
+#include <zephyr/drivers/flash/stm32_flash_api_extensions.h>
```
What about this?

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -64,6 +64,67 @@ struct flash_stm32_sector_t {
+	return (regs->OPTSR_CUR & FLASH_OPTSR_RDP_Msk) >> FLASH_OPTSR_RDP_Pos;
+}
+
+void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
```
Could you implement `write_optb()` function similar to what we have in `flash_stm32f4.c` and use it in this function? We plan to add write protection later.

duda-patryk replied:
The code is already duplicated. It doesn't matter if it's in `flash_stm32_set_rdp_level()` in `flash_stm32h7x.c` or in `write_optb()`. As I mentioned earlier we plan to add support for write protection in near future.

duda-patryk replied:
I see that you added `write_optb()` in the next commit. Can we squash `drivers: flash: stm32h7 flash has a write_optb function to set rdp level` with this commit?

duda-patryk reviewed (changes requested):
@FRASTM What do you think about @benediktibk suggestion about creating common file shared between both flash drivers, e.g. `flash_stm32_common.c`? Function `flash_stm32_option_bytes_lock()` seems to be the obvious candidate to be moved to that file.

FRASTM commented:
> @FRASTM What do you think about @benediktibk suggestion about creating common file shared between both flash drivers, e.g. `flash_stm32_common.c`? Function `flash_stm32_option_bytes_lock()` seems to be the obvious candidate to be moved to that file.

Sure, this makes sense, but should this bigger change be another iteration of _this_ PR

duda-patryk commented:
> > @FRASTM What do you think about @benediktibk suggestion about creating common file shared between both flash drivers, e.g. `flash_stm32_common.c`? Function `flash_stm32_option_bytes_lock()` seems to be the obvious candidate to be moved to that file.
> 
> Sure, this makes sense, but should this bigger change be another iteration of _this_ PR

Yeah, it should be done in a separate PR.

niedzwiecki-dawid commented:
@FRASTM have you had chance to see comments from @duda-patryk ? There is no update since his review and you requested a new review?

FRASTM commented:
Changed after review
Rebase on 1a5ae376a3f93312fd85016464a8b6221c554c14

FRASTM commented:
about @benediktibk suggestion about creating common file shared between both flash drivers
The common stm32 flash file is likely the existing flash_stm32_ex_op.c, containing the flash_stm32_option_bytes_lock() function. The flash_stm32_write_protection function becomes external, defined by the stm32 flash or stm32h7 flash driver.
See a proposal [here](https://github.com/FRASTM/zephyr/tree/stm32h7_rdp_common).

benediktibk commented:
> _about @benediktibk suggestion about creating common file shared between both flash drivers_
> 
> The common stm32 flash file is likely the existing flash_stm32_ex_op.c, containing the flash_stm32_option_bytes_lock() function. The flash_stm32_write_protection function becomes external, defined by the stm32 flash or stm32h7 flash driver. See a proposal [here](https://github.com/FRASTM/zephyr/tree/stm32h7_rdp_common).

Looks reasonable at the first glance.

FRASTM commented:
Rebase
drivers: flash: stm32h7 flash has a write_optb function to set rdp level squash into drivers: flash: stm32h7 with RDP protection
remove reference to H7 serie in the first change of flash_stm32.c
keep the #if defined(CONFIG_MULTITHREADING) || defined(CONFIG_STM32H7_DUAL_CORE)

benediktibk commented:
Minor issue, but there are still the copy paste left-overs for the licensing, see my comments above.

FRASTM commented:
> Minor issue, but there are still the copy paste left-overs for the licensing, see my comments above.

Sorry @benediktibk I cannot retrieve that, could you please point it ?

benediktibk commented:
The review comments are actually still open, they are about the *.overlay files in this PR. In there you still have SILA in the license information, probably a copy-paste leftover.

FRASTM commented:
> The review comments are actually still open, they are about the *.overlay files in this PR. In there you still have SILA in the license information, probably a copy-paste leftover.

Ok I see, correcting

benediktibk commented on tests/drivers/flash/stm32/boards/stm32h735g_disco.overlay:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2024 SILA Embedded Solutions GmbH
```
done

benediktibk commented on tests/drivers/flash/stm32/boards/stm32h745i_disco_stm32h745xx_m7.overlay:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2024 SILA Embedded Solutions GmbH
```
done

benediktibk commented on tests/drivers/flash/stm32/boards/stm32h750b_dk.overlay:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2024 SILA Embedded Solutions GmbH
```
done

niedzwiecki-dawid commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -62,32 +62,77 @@ struct flash_stm32_sector_t {
+		return rc;
+	}
+
+	regs->OPTCR = (regs->OPTCR & ~mask) | value;
```
Shouldn't `OPTSR_PRG` be used here? Sorry for late comment.

niedzwiecki-dawid replied:
@FRASTM ping :-)

MaazSk commented:
Sorry for late comment I tried the test on stm32h750b_dk but I got the following error. The error was `134` which means `Not supported`.


![image](https://github.com/user-attachments/assets/1e24cb00-1e57-4224-a8bc-97376d9b3bf7)
